source/extensions/common/matcher/domain_matcher.h

```diff
@@ -0,0 +1,197 @@
+    parent_prefix_ = parent_prefix;
+    if (prefix_on_match_) {
+      parent_prefix = makeOptRef(*this);
+    }
+    for (auto& [_, child] : children_) {
+      child.link(parent_prefix);
+    }
+  }
+
+  // Find the deepest node matching the reversed parts of the domain.
+  // Returns a node and a bool indicating whether the match is exact for this node.
+  std::pair<DomainNode const*, bool> find(absl::Span<const std::string> parts) const {
+    if (parts.empty()) {
+      return std::make_pair(this, true);
+    }
+    auto it = children_.find(parts[0]);
+    if (it != children_.end()) {
+      return it->second.find(parts.subspan(1));
+    }
+    return std::make_pair(this, false);
+  }
+};
+
+template <class DataType> struct DomainTree {
+  DomainNode<DataType> root_;
+  std::vector<OnMatch<DataType>> on_matches_;
+};
+
+/**
+ * General utilities for domain name matching.
+ */
+class DomainMatcherUtility {
+public:
+  static void validateServerName(const std::string& server_name);
+  static void duplicateServerNameError(const std::string& server_name);
+};
+
+/**
+ * Implementation of a `sublinear` domain matcher using a trie.
+ **/
+template <class DataType> class DomainMatcher : public MatchTree<DataType> {
+public:
+  DomainMatcher(DataInputPtr<DataType>&& data_input, absl::optional<OnMatch<DataType>> on_no_match,
+                const std::shared_ptr<DomainTree<DataType>>& domain_tree)
+      : data_input_(std::move(data_input)), on_no_match_(std::move(on_no_match)),
+        domain_tree_(domain_tree) {}
+
+  typename MatchTree<DataType>::MatchResult match(const DataType& data) override {
+    const auto input = data_input_->get(data);
+    if (input.data_availability_ != DataInputGetResult::DataAvailability::AllDataAvailable) {
+      return {MatchState::UnableToMatch, absl::nullopt};
+    }
+    if (!input.data_) {
+      return {MatchState::MatchComplete, on_no_match_};
+    }
+    std::vector<std::string> parts = absl::StrSplit(*input.data_, ".");
+    std::reverse(parts.begin(), parts.end());
+    // Traverse from the most specific match to the root and check for prefix matches.
+    auto [node, exact] = domain_tree_->root_.find(parts);
+    while (node) {
+      OptRef<OnMatch<DataType>> on_match;
+      if (exact) {
+        on_match = node->exact_on_match_;
+        exact = false;
+      } else {
+        on_match = node->prefix_on_match_;
+      }
+      if (on_match) {
+        if (on_match->action_cb_) {
+          return {MatchState::MatchComplete, OnMatch<DataType>{on_match->action_cb_, nullptr}};
+        }
+        auto matched = evaluateMatch(*on_match->matcher_, data);
+        if (matched.match_state_ == MatchState::UnableToMatch) {
+          return {MatchState::UnableToMatch, absl::nullopt};
+        }
+        if (matched.match_state_ == MatchState::MatchComplete && matched.result_) {
+          return {MatchState::MatchComplete, OnMatch<DataType>{matched.result_, nullptr}};
+        }
```

Comment:
Is it possible to escape the `if (on_match) {` block w/o `return` ?
In which case?

If the answer is yes, should on_match degrade from exact_on_match to prefix_on_match ?

Reply:
Yes, it can have nested matchers so the match can complete without a result.
It does degrade, because it traverses the `parent_prefix_` edges. E.g. it'll try wider and wider patters by going from the node to the root.